File: DataStructures/sumPerfectSquare.js
const SumsPerfectSquare = num => {
  let squares = [];
  let all = [];
  let sum = 0;
  for (let i = 1; i <= Math.sqrt(num); i++) {
    squares.push(i * i);
  }
  for (let sq of squares) {
    sum += sq;
    all.push(sq);
  }
  return squares;
};
console.log(SumsPerfectSquare(15));

File: SolutionsToProblems/sumPerfectSquare.py
# def getMinSquares(n):
#     if n <= 3:
#         return n
#     res = n
#     for x in range(1, n+1):
#         temp = x * x
#         if temp > n:
#             break
#         else:
#             res = min(res, 1 + getMinSquares(n - temp))

#     return res


# print(getMinSquares(15))
import math


def numSquares(n):
    if n <= 3:
        return n
    squares = {i * i for i in range(1, int(math.floor(math.sqrt(n))) + 1)}
    sums = squares
    for i in range(1, n):
        if n in sums:
            return i
        sums = {a + b for a in squares for b in sums if a + b <= n}
        print(sums)
    return sums


print(numSquares(15))
